builtin/process_osh.py

```diff
@@ -682,4 +684,119 @@ def Run(self, cmd_val):
         return 0
 
 
+class Kill(vm._Builtin):
+    """Send a signal to a process"""
+    def __init__(self, job_list):
+        # type: (process.JobList) -> None
+        self.job_list = job_list
+
+    def _SignameToSignum(self, name):
+        # type: (str) -> int
+        signal_name = name.upper()
+        if signal_name.startswith('SIG'):
+            signal_name = signal_name[3:]
+        return signal_def.GetNumber(signal_name)
+
+    def _ParsePid(self, pid_arg, pid_arg_loc):
+        # type: (str, loc_t) -> int
+        if pid_arg.startswith('%'):
+            job = self.job_list.JobFromSpec(pid_arg)
+            if job is None:
+                e_usage('invalid signal specification %r' % pid_arg, pid_arg_loc)
+            else:
+                return job.ProcessGroupId()
+        else:
+            try:
+                target_pid = int(pid_arg)
+                return target_pid
+            except ValueError:
+                e_usage('invalid process id specification %r' % pid_arg, pid_arg_loc)
+    
+    def Run(self, cmd_val):
+        # type: (cmd_value.Argv) -> int
+
+        signal_to_send = signal_def.NO_SIGNAL
+        target_pid = 0
+
+        arg_r = args.Reader(cmd_val.argv, locs=cmd_val.arg_locs)
+        arg_r.Next()
+        sigspec_arg, sigspec_arg_loc = arg_r.Peek2()
+        if len(cmd_val.argv) == 3 and sigspec_arg.startswith('-'):
+            # if we land in this if or elif statement,
+            # we are dealing with the -sigspec argument, which
+            # flag-util.ParseCmdVal cant deal with so we parse it manually
+            if(sigspec_arg.startswith('-')):
+                if sigspec_arg[1:].isdigit():
+                    signal_to_send = int(sigspec_arg[1:])
+                    if(signal_to_send < 0):
+                        e_usage('invalid signal specification %r' % sigspec_arg, sigspec_arg_loc)
+                elif len(sigspec_arg[1:]) > 2:
+                    signal_to_send = self._SignameToSignum(sigspec_arg[1:])
+                    if(signal_to_send < 0):
+                        e_usage('invalid signal specification %r' % sigspec_arg, sigspec_arg_loc)
+
+            if signal_to_send != signal_def.NO_SIGNAL:
+                arg_r.Next()
+                pid_arg, pid_arg_loc = arg_r.Peek2()
+                target_pid = self._ParsePid(pid_arg, pid_arg_loc)
+
+        # we didn't have to deal with a -sigspec argument, so we can parse as usual
+        if signal_to_send == signal_def.NO_SIGNAL:
```

Comment:
I think we could return early for the `kill -TERM 123` case, so then we don't have to indent this `kill -l` logic under `if signal_to_send ...`

I like the early return style for the different usages (similar to the `trap` builtin)